### ISSUE TITLE
algebra naming convention 

### ISSUE THREAD

**Issue by fnrizzi** [Pressio/pressio#80]
About to rethink how certain operations are called. So I like to have a voting on this :) 

#
Design A - we define separate operations for matrices and multivectors: 
- **MATRIX** operations: 
Since for ROMs we are dealing with dense operations, I would suggest keeping BLAS-like names as. 
1. matrix-matrix product `C = beta*C + alpha*op(A)*op(B)`
will be `pressio::containers::ops::gemm(modeA, modeB, alpha, A, B, beta, C)` (consistent with Blas and Kokkos).  Where `op(A)` means `A` or `A^T`.  
We can have: `modeA = N, T` like in original Blas or use a more expressive tag like: `modeA = pressio::transpose()` and `modeA = pressio::nontranspose()`
2. matrix vector product `y = beta * y + alpha * op(A) * x`
will be `pressio::containers::ops::gemv(modeA, alpha, A, x, beta, y);` (consistent with Blas and Kokkos). 

- **MULTI VECTOR** operations: 
In pressio we also have the concept of multivector (mv). 
1. mv-mv product `C = beta*C + alpha*op(A)*op(B)`
will be `pressio::containers::ops::gemvmv(modeA, modeB, alpha, A, B, beta, C)`. 
1. mv-vec product `y = beta*y + alpha*op(A)*x`
will be `pressio::containers::ops::gemvv(modeA, alpha, A, x, beta, y)`. 

#
Design B - we use same naming for matrices and multivectors: 
Regardless of matrix or mv, we call everything as : 
- `pressio::containers::ops::gemm(modeA, modeB, alpha, A, B, beta, C)`
- `pressio::containers::ops::gemv(modeA, alpha, A, x, beta, y)`

(it is implicitly intended that we specialize (if needed) gemm and gemv based on whether operands are matrices or mv.)

**Comment by fnrizzi** [comment]
@eparish1 @pjb236 @mhoemmen @jtencer @vbrunini can you please vote for design A or B and/or share any suggestions? Thanks

**Comment by pjb236** [comment]
I vote for design B.

**Comment by mhoemmen** [comment]
This is a fun debate! :-)  My view is that concrete operations deserve more concrete names, and abstract operations call for more abstract names.  For example, "`gemm`" means "dense matrix represented in the General format, times a dense matrix."  If you can use the same interface to multiply a sparse matrix times a dense matrix, then "`gemm`" is not the right name.  "`mm`" or "`matrix_matrix_multiply`" might be.  The BLAS uses the names that it does because the BLAS routines and functions are not polymorphic on the matrix storage format.  The Fortran 9x bindings _are_ polymorphic on the matrix element type, so they omit the element type prefix (S, D, C, or Z).  If we want an interface that's polymorphic on the matrix storage format as well, then we should analogously omit the matrix storage format prefix (e.g., "GE").

Regarding whether the interface should look different for vectors vs. multivectors, I think this depends on how abstract you mean the interface to be.  The [Sparse BLAS](http://www.netlib.org/blas/blast-forum/chapter3.pdf) provides operations for sparse matrices, not for generic "operators."  Likely not coincidentally, it looks more like Design B.  For instance, the routine that multiplies a sparse matrix with a dense matrix ("multivector") is called "MM" (matrix-matrix multiply) with a prefix "US" (unstructured sparse) instead of "GE" (general dense).  The routine that multiplies a sparse matrix with a dense vector is called "MV."

Trilinos takes a more generic approach that gives vectors and multivectors the same interface.  `Operator::apply` (the sparse matrix-vector multiply method) takes MultiVectors as input and output.  Note that Trilinos chose the name "apply," not "gemv," "gemvmv," or "gemm."  Trilinos is not a BLAS; its model of computation is based on abstract functions between vector spaces.  `Operator::apply` could apply a preconditioner or even a nonlinear operator.

In summary, I think if you mean the interface to be abstract and to behave like Trilinos' `Operator::apply`, then an abstract name like "`apply`" is more appropriate, and could be the same for both dense vectors and dense multivectors.  If you mean the interface to work on concrete matrices, not abstract operators, then BLAS-like names are more appropriate.

**Comment by fnrizzi** [comment]
@pjb236 and @mhoemmen thanks! 
This is a fun debate, indeed :) 
The reason I brought this up is that @eparish1 brought this up to me, so it is a good time to finalize the interface once for good! 

If we stay somewhat general, we could have: 
1. `ops::mvproduct()` where: 
    * `m` = matrix (of any kind) or multivector 
    * `v`= a dense vector. 
2. `ops::mmproduct()` where: 
    * `m` = matrix (of any kind) or multivector 

Or we can push the limits and collapse everything into: `ops::product()` for everything (since we can detect what types are passed in).

**Comment by mhoemmen** [comment]
Collapsing mv and mm into "product" makes sense.  To me, the more important thing is not to make vector a subclass of multivector.  One of the symptoms of this mistake is that Belos conflates "block" (in the sense of block Krylov methods) with "collection of blocks."  For instance, Belos treats a MultiVector as a memory pool for either single vectors or multivectors (for block Krylov methods).  This forces MultiVector to have a column-major layout on each process, which is bad for performance.  In order to get some of that performance back, kokkos-kernels needs a more complicated local sparse matrix-multivector multiply kernel that takes longer to build.

**Comment by fnrizzi** [comment]
> Collapsing mv and mm into "product" makes sense. 

Great! 

> To me, the more important thing is not to make vector a subclass of multivector. 

Not our case... these are two distinct types. 

> One of the symptoms of this mistake is that Belos conflates "block" (in the sense of block Krylov methods) with "collection of blocks." For instance, Belos treats a MultiVector as a memory pool for either single vectors or multivectors (for block Krylov methods). This forces MultiVector to have a column-major layout on each process, which is bad for performance. In order to get some of that performance back, kokkos-kernels needs a more complicated local sparse matrix-multivector multiply kernel that takes longer to build.

Thanks for pointing out this case!
When you say "forces MV to have column-major layout, which is bad for performance" you are referring to the Krylov methods specifically, right?

**Comment by mhoemmen** [comment]
@fnrizzi wrote:
> When you say "forces MV to have column-major layout, which is bad for performance" you are referring to the Krylov methods specifically, right?

Yes, that's right.  For sparse matrices stored in compressed sparse row format, it's easier to write a fast kernel for sparse matrix - multivector multiply if the multivector is row major than column major.

Anasazi and Belos conflate "block" with "memory pool of blocks," so they can't easily be changed to work with row-major multivectors.  Furthermore, MultiVector's interface itself assumes that a column of a MultiVector is a Vector.

**Comment by mhoemmen** [comment]
@fnrizzi wrote:
> When you say "forces MV to have column-major layout, which is bad for performance" you are referring to the Krylov methods specifically, right?

Yes, that's right.  For sparse matrices stored in compressed sparse row format, it's easier to write a fast kernel for sparse matrix - multivector multiply if the multivector is row major than column major.

Anasazi and Belos conflate "block" with "memory pool of blocks," so they can't easily be changed to work with row-major multivectors.  Furthermore, MultiVector's interface itself assumes that a column of a MultiVector is a Vector.
@fnrizzi wrote:
> > To me, the more important thing is not to make vector a subclass of multivector.
> 
> Not our case... these are two distinct types.

You did it right :-D

**Comment by eparish1** [comment]
Sorry to be late to the party! From my perspective, collapsing into ops::product makes a lot of sense to me from the users point of view. I am of the opinion that, as a user, having a common interface for basic matrix-vector operations -- regardless of the underlying type so long as they meet certain requirements -- is the most intuitive. I think this will be helpful down the line as well for minimal code maintenance. I will say, however, that I don't have as good of an appreciation unforeseen implications this may result in as others on this thread do.

**Comment by fnrizzi** [comment]
The solution will look like this: 

1. `y = beta * y + alpha * op(A) * x` where
   * `y`: is `pressio::containers::Vector`
   * `A`: is `pressio::containers::Matrix` or `pressio::containers::Multivector`

can be called with: 
```c++
pressio::containers::ops::product(modeA, alpha, A, x, beta, y)
```

2. `C = beta * C + alpha * op(A) * op(B)` where
   * `C`: is `pressio::containers::Matrix` or `pressio::containers::Multivector`
   * `A`: is `pressio::containers::Matrix` or `pressio::containers::Multivector`
   * `B`: is `pressio::containers::Matrix` or `pressio::containers::Multivector`

can be called with: 
```c++
pressio::containers::ops::product(modeA, modeB, alpha, A, B, beta, C)
```

In all cases, we have: 
```c++
op(G) = G^T : use mode = pressio::transpose() 
op(G) = G   : use mode = pressio::nontranspose() 
```

**Comment by mhoemmen** [comment]
@fnrizzi FYI, {E,T}petra do not implement (sparse matrix)*transpose(X).  I could imagine this being a useful operation, but there's no way to do it with the current interface.

Details: The "column Map" of a MultiVector in {E,T}petra is always locally replicated.  (Rather, there is no explicit column Map for MultiVectors in {E,T}petra, but if there were, it would be locally replicated.  This is a bit different in Thyra, where MultiVectors are linear operators and therefore have explicit Maps.)  This means there's no way to represent a MultiVector whose transpose has globally distributed rows.

**Comment by fnrizzi** [comment]
> @fnrizzi FYI, {E,T}petra do not implement (sparse matrix)*transpose(X). I could imagine this being a useful operation, but there's no way to do it with the current interface.

Yes, but fortunately we only care for dense operations (for now) :) 

> 
> Details: The "column Map" of a MultiVector in {E,T}petra is always locally replicated. (Rather, there is no explicit column Map for MultiVectors in {E,T}petra, but if there were, it would be locally replicated. This is a bit different in Thyra, where MultiVectors are linear operators and therefore have explicit Maps.) This means there's no way to represent a MultiVector whose transpose has globally distributed rows.

Yes, i remember that! Thanks for clarifying!

**Comment by fnrizzi** [comment]
bf54c7b28e3841ffddba9bf353214add7e359ab0
